File: cpp/carfac_test.cc
// Copyright 2013 The CARFAC Authors. All Rights Reserved.
// Author: Alex Brandmeyer
//
// This file is part of an implementation of Lyon's cochlear model:
// "Cascade of Asymmetric Resonators with Fast-Acting Compression"
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "carfac.h"

#include <cmath>
#include <string>
#include <vector>

#include "benchmark/benchmark.h"
#include "gtest/gtest.h"
#include "agc.h"
#include "car.h"
#include "common.h"
#include "ear.h"
#include "ihc.h"
#include "test_util.h"
#include <Eigen/Core>

// Reads a two dimensional vector of audio data from a text file
// containing the output of the Matlab wavread() function.
ArrayXX LoadAudio(const std::string& filename, int num_samples, int num_ears) {
  // The Matlab audio input is transposed compared to the C++.
  return LoadMatrix(filename, num_samples, num_ears).transpose();
}

// Writes the CARFAC NAP output to a text file.
void WriteNAPOutput(const CARFACOutput& output, const std::string& filename,
                    int ear) {
  WriteMatrix(filename, output.nap()[ear].transpose());
}

TEST(CarTest, PoleFrequencies) {
  constexpr FPType SampleRateHz = 16000;
  const CARParams params;
  ArrayX poles = CARPoleFrequencies(SampleRateHz, params);
  for (int pole_index = 0; pole_index < poles.size(); ++pole_index) {
    SCOPED_TRACE(std::string("pole_index: ") +
                 testing::PrintToString(pole_index));
    const FPType pole = poles[pole_index];
    ASSERT_NEAR(pole,
                CARChannelIndexToFrequency(SampleRateHz, params, pole_index),
                1e-5f * pole);
    ASSERT_NEAR(pole_index,
                CARFrequencyToChannelIndex(SampleRateHz, params, pole),
                1e-5f * pole_index);
  }
}

class CARFACTest : public testing::Test {
 protected:
  CARFACTest() : open_loop_(false) {}

  std::vector<ArrayXX> LoadTestData(const std::string& basename,
                                    int num_samples,
                                    int num_ears,
                                    int num_channels) const {
    std::vector<ArrayXX> test_data;
    for (int ear = 0; ear < num_ears; ++ear) {
      std::string filename = basename + std::to_string(ear + 1) + ".txt";
      // The Matlab CARFAC output is transposed compared to the C++.
      test_data.push_back(
          LoadMatrix(filename, num_samples, num_channels).transpose());
    }
    return test_data;
  }

  void AssertCARFACOutputNear(const std::vector<ArrayXX>& expected,
                              const std::vector<ArrayXX>& actual) const {
    ASSERT_EQ(expected.size(), actual.size());
    for (int ear = 0; ear < expected.size(); ++ear) {
      AssertArrayNear(expected[ear], actual[ear], kTestPrecision);
    }
  }

  void RunCARFACAndCompareWithMatlab(const std::string& test_name,
                                     int num_samples,
                                     int num_ears,
                                     int num_channels,
                                     FPType sample_rate) const {
    ArrayXX sound_data =
        LoadAudio(test_name + "-audio.txt", num_samples, num_ears);

    CARFAC carfac(num_ears, sample_rate, car_params_, ihc_params_, agc_params_);
    CARFACOutput output(true, true, false, false);
    carfac.RunSegment(sound_data, open_loop_, &output);

    std::vector<ArrayXX> expected_nap = LoadTestData(
        test_name + "-matlab-nap", num_samples, num_ears, num_channels);
    AssertCARFACOutputNear(expected_nap, output.nap());
    std::vector<ArrayXX> expected_bm = LoadTestData(
        test_name + "-matlab-bm", num_samples, num_ears, num_channels);
    AssertCARFACOutputNear(expected_bm, output.bm());

    // TODO(dicklyon): Conditionally overwrite test output files.
    // WriteNAPOutput(output, test_name + "-cpp-nap1.txt", 0);
    // WriteNAPOutput(output, test_name + "-cpp-nap2.txt", 1);
  }

  CARParams car_params_;
  IHCParams ihc_params_;
  AGCParams agc_params_;
  bool open_loop_;
};

TEST_F(CARFACTest, MatchesMatlabOnBinauralData) {
  const int kNumSamples = 882;
  const int kNumEars = 2;
  const int kNumChannels = 71;
  const FPType kSampleRate = 22050.0;
  RunCARFACAndCompareWithMatlab(
      "binaural_test", kNumSamples, kNumEars, kNumChannels, kSampleRate);
}

TEST_F(CARFACTest, MatchesMatlabOnLongBinauralData) {
  const int kNumSamples = 2000;
  const int kNumEars = 2;
  const int kNumChannels = 83;
  const FPType kSampleRate = 44100.0;
  RunCARFACAndCompareWithMatlab(
      "long_test", kNumSamples, kNumEars, kNumChannels, kSampleRate);
}

TEST_F(CARFACTest, CanDisableAGC) {
  const FPType kSampleRate = 8000.0;  // Hz.
  const int kNumEars = 1;
  const int kNumSamples = static_cast<int>(kSampleRate);

  // Sinusoid input.
  const float kFrequency = 10.0;  // Hz.
  ArrayXX sound_data(kNumEars, kNumSamples);
  sound_data.row(0) =
      ArrayX::LinSpaced(kNumSamples, 0.0, 2 * kFrequency * M_PI).sin();

  CARFACOutput output(false, false, false, true);  // store_agc

  // AGC enabled.
  CARFAC carfac(kNumEars, kSampleRate, car_params_, ihc_params_, agc_params_);
  const bool kOpenLoop = false;
  carfac.RunSegment(sound_data, kOpenLoop, &output);
  ArrayXX agc_enabled_output = output.agc()[0];

  // AGC disabled.
  agc_params_.num_stages = 0;
  carfac.Redesign(kNumEars, kSampleRate, car_params_, ihc_params_, agc_params_);
  carfac.RunSegment(sound_data, kOpenLoop, &output);
  ArrayXX agc_disabled_output = output.agc()[0];

  ASSERT_EQ(agc_enabled_output.size(), agc_disabled_output.size());
  EXPECT_TRUE((agc_enabled_output !=  agc_disabled_output.size()).any());
  for (int c = 0; c < agc_disabled_output.rows(); ++c) {
    // With the AGC disabled, the agc output for a given channel
    // should be identical at all times.
    EXPECT_TRUE(
        (agc_disabled_output.row(c) == agc_disabled_output(c, 0)).all());
    EXPECT_FALSE(
        (agc_enabled_output.row(c) == agc_enabled_output(c, 0)).all());
  }
}

TEST_F(CARFACTest, PoleFrequenciesAreDecreasing) {
  const FPType kSampleRate = 8000.0;  // Hz.
  const int kNumEars = 1;

  CARFAC carfac(kNumEars, kSampleRate, car_params_, ihc_params_, agc_params_);
  ArrayX pole_frequencies = carfac.pole_frequencies();
  EXPECT_EQ(carfac.num_channels(), pole_frequencies.size());
  for (int i = 1; i < pole_frequencies.size(); ++i) {
    EXPECT_LT(pole_frequencies(i), pole_frequencies(i - 1));
  }
}

TEST_F(CARFACTest, MatchesMatlabWithAGCOff) {
  const int kNumSamples = 2000;
  const int kNumEars = 2;
  const int kNumChannels = 83;
  const FPType kSampleRate = 44100.0;
  open_loop_ = true;  // Turns off the AGC feedback.
  RunCARFACAndCompareWithMatlab(
      "agc_test", kNumSamples, kNumEars, kNumChannels, kSampleRate);
}

TEST_F(CARFACTest, MatchesMatlabWithIHCJustHalfWaveRectifyOn) {
  const int kNumSamples = 2000;
  const int kNumEars = 2;
  const int kNumChannels = 83;
  const FPType kSampleRate = 44100.0;
  ihc_params_.just_half_wave_rectify = true;
  RunCARFACAndCompareWithMatlab(
      "ihc_just_hwr_test", kNumSamples, kNumEars, kNumChannels, kSampleRate);
}

TEST_F(CARFACTest, AGCDesignBehavesSensibly) {
  const int kNumEars = 2;
  const FPType kSampleRate = 22050.0;
  const std::vector<FPType> spreads = {0.0, 1.0, 1.4, 2.0, 2.8, 3.5, 4.0};
  for (int i = 0; i < spreads.size(); ++i) {
    FPType spread = spreads[i];
    EXPECT_GE(spread, 0);
    // Test zero and default and larger AGC spread.
    agc_params_.agc1_scales[0] = 1.0 * spread;
    agc_params_.agc2_scales[0] = 1.65 * spread;
    for (int i = 1; i < agc_params_.agc1_scales.size(); ++i) {
      agc_params_.agc1_scales[i] = agc_params_.agc1_scales[i - 1] * sqrt(2.0);
      agc_params_.agc2_scales[i] = agc_params_.agc2_scales[i - 1] * sqrt(2.0);
    }
    CARFAC carfac(kNumEars, kSampleRate, car_params_, ihc_params_, agc_params_);
    for (int stage = 0; stage < 4; ++stage) {
      // Stages all should be alike, as their time constants quadruple, sample
      // rates halve, and variances double at each stage, so equivalent filters.
      const AGCCoeffs& coeffs = carfac.get_ear(0).get_agc_coeffs(stage);
      switch (i) {
        case 0:
          EXPECT_EQ(3, coeffs.agc_spatial_n_taps);
          EXPECT_EQ(0, coeffs.agc_spatial_iterations);
          break;
        case 1:
          EXPECT_EQ(3, coeffs.agc_spatial_n_taps);
          EXPECT_EQ(1, coeffs.agc_spatial_iterations);
          break;
        case 2:
          EXPECT_EQ(5, coeffs.agc_spatial_n_taps);
          EXPECT_EQ(1, coeffs.agc_spatial_iterations);
          break;
        case 3:
          EXPECT_EQ(5, coeffs.agc_spatial_n_taps);
          EXPECT_EQ(2, coeffs.agc_spatial_iterations);
          break;
        case 4:
          EXPECT_EQ(5, coeffs.agc_spatial_n_taps);
          EXPECT_EQ(3, coeffs.agc_spatial_iterations);
          break;
        case 5:
          EXPECT_EQ(5, coeffs.agc_spatial_n_taps);
          EXPECT_EQ(4, coeffs.agc_spatial_iterations);
          break;
        case 6:
          EXPECT_EQ(-1, coeffs.agc_spatial_iterations);
          break;
      }
    }
  }
}

TEST_F(CARFACTest, AGCDesignAtLowSampleRate) {
  const int kNumEars = 2;
  const FPType kSampleRate = 8000.0;
  const std::vector<FPType> spreads = {0.0, 1.0, 1.4, 2.0, 2.8};
  for (int i = 0; i < spreads.size(); ++i) {
    FPType spread = spreads[i];
    EXPECT_GE(spread, 0);
    // Test zero and default and larger AGC spread.

    agc_params_.agc1_scales[0] = 1.0 * spread;
    agc_params_.agc2_scales[0] = 1.65 * spread;
    for (int i = 1; i < agc_params_.agc1_scales.size(); ++i) {
      agc_params_.agc1_scales[i] = agc_params_.agc1_scales[i - 1] * sqrt(2.0);
      agc_params_.agc2_scales[i] = agc_params_.agc2_scales[i - 1] * sqrt(2.0);
    }
    CARFAC carfac(kNumEars, kSampleRate, car_params_, ihc_params_, agc_params_);
    const AGCCoeffs& coeffs = carfac.get_ear(0).get_agc_coeffs(0);
    switch (i) {
      case 0:
        EXPECT_EQ(3, coeffs.agc_spatial_n_taps);
        EXPECT_EQ(0, coeffs.agc_spatial_iterations);
        break;
      case 1:
        EXPECT_EQ(5, coeffs.agc_spatial_n_taps);
        EXPECT_EQ(1, coeffs.agc_spatial_iterations);
        break;
      case 2:
        EXPECT_EQ(5, coeffs.agc_spatial_n_taps);
        EXPECT_EQ(2, coeffs.agc_spatial_iterations);
        break;
      case 3:
        EXPECT_EQ(5, coeffs.agc_spatial_n_taps);
        EXPECT_EQ(4, coeffs.agc_spatial_iterations);
        break;
      case 4:
        EXPECT_EQ(-1, coeffs.agc_spatial_iterations);
        break;
    }
  }
}

void BM_CarfacSegment(benchmark::State& state) {
  const auto segment_length_samples = state.range(0);
  const int num_ears = 1;
  const FPType sample_rate = 22050.0;
  CARParams car_params;
  IHCParams ihc_params;
  AGCParams agc_params;
  CARFAC carfac(num_ears, sample_rate, car_params, ihc_params, agc_params);
  // Sinusoid input.
  const float kFrequency = 500.0;  // Hz.
  ArrayXX sound_data(num_ears, segment_length_samples);
  sound_data.row(0) =
      ArrayX::LinSpaced(segment_length_samples, 0.0, 2 * kFrequency * M_PI)
          .sin();

  const bool open_loop = false;
  for (auto s : state) {
    // store everything for the benchmarks.
    CARFACOutput output(true, true, true, true);
    carfac.RunSegment(sound_data, open_loop, &output);
  }
}

BENCHMARK(BM_CarfacSegment)->ThreadRange(1, 128)
    ->Arg(220)
    ->Arg(2205)
    ->Arg(22050)
    ->Arg(44100)
    ->Arg(220500);

File: python/tf/carfac_test.py
# Lint as: python3
#!/usr/bin/env python

# Copyright 2021 The CARFAC Authors. All Rights Reserved.
#
# This file is part of an implementation of Lyon's cochlear model:
# "Cascade of Asymmetric Resonators with Fast-Acting Compression"
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Tests for carfac.python.tf.carfac."""

from typing import Optional

import numpy as np
from parameterized import parameterized
import tensorflow as tf

from .. import testing as carfac_testing
from . import carfac


class _TestCallable:

  def __init__(self, cell: carfac.CARFACCell):
    cell.call = tf.function(cell.call)
    self.cell = cell
    self.layer = tf.function(tf.keras.layers.RNN(
        self.cell, return_sequences=True))

  def __call__(self, audio: np.ndarray) -> np.ndarray:
    return self.layer(np.reshape(audio, (1,
                                         audio.shape[0],
                                         audio.shape[1],
                                         1))).numpy()[0]


class CARFACTest(tf.test.TestCase):

  def testConvolvers(self):
    # Verifies that all convolver options produce the same values.
    for kernel_size in [3, 5]:
      previous_values: Optional[tf.Tensor] = None
      kernel = tf.ones((kernel_size,), dtype=tf.float64) * 1.3
      data = (tf.ones((3, 4, 92), dtype=tf.float64) *
              tf.range(92, dtype=tf.float64))
      for convolver in carfac.convolution_methods.values():
        output = convolver(tf.constant(5, dtype=tf.int32), kernel, data)
        if previous_values is None:
          previous_values = output
        else:
          np.testing.assert_allclose(previous_values, output)

  def testRecurrenceExpanders(self):
    # Verifies that all recurrence expanders produce the same values.
    previous_value: Optional[tf.Tensor] = None
    a_0 = (tf.ones((3, 4), dtype=tf.float64) *
           tf.constant(0.51, dtype=tf.float64))
    f = (0.9 + (0.2 / 92) * tf.ones((3, 4, 92), dtype=tf.float64) *
         tf.range(92, dtype=tf.float64))
    g = (0.01 + (2.0 / 92) * tf.ones((3, 4, 92), dtype=tf.float64) *
         tf.range(92, dtype=tf.float64))
    for expander in carfac.recurrence_expansion_methods.values():
      output = expander(a_0, f, g)
      if previous_value is None:
        previous_value = output
      else:
        np.testing.assert_allclose(previous_value, output)

  def testMatchesMatlabWithAGCOff(self):
    carfac_testing.assert_matlab_compatibility(
        'agc_test', _TestCallable(
            carfac.CARFACCell(
                open_loop=True,
                num_ears=2,
                outputs=[carfac.CARFACOutput.BM,
                         carfac.CARFACOutput.NAP],
                car_params=carfac.CARParams(
                    sample_rate_hz=tf.constant(44100.0)))))

  def testMatchesMatlabOnBinauralData(self):
    carfac_testing.assert_matlab_compatibility(
        'binaural_test', _TestCallable(
            carfac.CARFACCell(
                num_ears=2,
                outputs=[carfac.CARFACOutput.BM,
                         carfac.CARFACOutput.NAP],
                car_params=carfac.CARParams(
                    sample_rate_hz=tf.constant(22050.0)))))

  def testMatchesMatlabOnLongBinauralData(self):
    carfac_testing.assert_matlab_compatibility(
        'long_test', _TestCallable(
            carfac.CARFACCell(
                num_ears=2,
                outputs=[carfac.CARFACOutput.BM,
                         carfac.CARFACOutput.NAP],
                car_params=carfac.CARParams(
                    sample_rate_hz=tf.constant(44100.0)))))

  def testMatchesMatlabWithIHCJustHalfWaveRectifyOn(self):
    carfac_testing.assert_matlab_compatibility(
        'ihc_just_hwr_test', _TestCallable(
            carfac.CARFACCell(
                num_ears=2,
                ihc_params=carfac.IHCParams(
                    just_half_wave_rectify=tf.constant(1.0)),
                outputs=[carfac.CARFACOutput.BM,
                         carfac.CARFACOutput.NAP],
                car_params=carfac.CARParams(
                    sample_rate_hz=tf.constant(44100.0)))))

  def testAGCDesignAtLowSampleRate(self):
    for i, spread in enumerate([0.0, 1.0, 1.4, 2.0, 2.8]):
      carfac_cell = carfac.CARFACCell(
          car_params=carfac.CARParams(sample_rate_hz=tf.constant(8000.0)),
          agc_params=carfac.AGCParams(agc1_scales0=tf.constant(1.0 * spread),
                                      agc2_scales0=tf.constant(1.65 * spread)))
      agc_coeffs = carfac_cell._design_agc_coeffs()
      if i == 0:
        self.assertTrue(
            np.all(agc_coeffs.agc_spatial_n_taps.numpy() == 3))
        self.assertTrue(
            np.all(agc_coeffs.agc_spatial_iterations.numpy() == 0))
      elif i == 1:
        self.assertTrue(
            np.all(agc_coeffs.agc_spatial_n_taps.numpy() == 5))
        self.assertTrue(
            np.all(agc_coeffs.agc_spatial_iterations.numpy() == 1))
      elif i == 2:
        self.assertTrue(
            np.all(agc_coeffs.agc_spatial_n_taps.numpy() == 5))
        self.assertTrue(
            np.all(agc_coeffs.agc_spatial_iterations.numpy() == 2))
      elif i == 3:
        self.assertTrue(
            np.all(agc_coeffs.agc_spatial_n_taps.numpy() == 5))
        self.assertTrue(
            np.all(agc_coeffs.agc_spatial_iterations.numpy() == 4))
      elif i == 4:
        self.assertTrue(
            np.all(
                agc_coeffs.agc_spatial_iterations.numpy() == -1))

  def testAGCDesignBehavesSensibly(self):
    for i, spread in enumerate([0.0, 1.0, 1.4, 2.0, 2.8, 3.5, 4.0]):
      carfac_cell = carfac.CARFACCell(
          car_params=carfac.CARParams(sample_rate_hz=tf.constant(22050.0)),
          agc_params=carfac.AGCParams(agc1_scales0=tf.constant(1.0 * spread),
                                      agc2_scales0=tf.constant(1.65 * spread)))
      agc_coeffs = carfac_cell._design_agc_coeffs()
      if i == 0:
        self.assertTrue(
            np.all(agc_coeffs.agc_spatial_n_taps.numpy() == 3))
        self.assertTrue(
            np.all(agc_coeffs.agc_spatial_iterations.numpy() == 0))
      elif i == 1:
        self.assertTrue(
            np.all(agc_coeffs.agc_spatial_n_taps.numpy() == 3))
        self.assertTrue(
            np.all(agc_coeffs.agc_spatial_iterations.numpy() == 1))
      elif i == 2:
        self.assertTrue(
            np.all(agc_coeffs.agc_spatial_n_taps.numpy() == 5))
        self.assertTrue(
            np.all(agc_coeffs.agc_spatial_iterations.numpy() == 1))
      elif i == 3:
        self.assertTrue(
            np.all(agc_coeffs.agc_spatial_n_taps.numpy() == 5))
        self.assertTrue(
            np.all(agc_coeffs.agc_spatial_iterations.numpy() == 2))
      elif i == 4:
        self.assertTrue(
            np.all(agc_coeffs.agc_spatial_n_taps.numpy() == 5))
        self.assertTrue(
            np.all(agc_coeffs.agc_spatial_iterations.numpy() == 3))
      elif i == 5:
        self.assertTrue(
            np.all(agc_coeffs.agc_spatial_n_taps.numpy() == 5))
        self.assertTrue(
            np.all(agc_coeffs.agc_spatial_iterations.numpy() == 4))
      elif i == 6:
        self.assertTrue(
            np.all(
                agc_coeffs.agc_spatial_iterations.numpy() == -1))
    pass

  @parameterized.expand((
      (1.0, 1.0, True, tf.float32),
      (1.0, 1.0, False, tf.float32),
      (1.0, 0.0, True, tf.float32),
      (1.0, 0.0, False, tf.float32),
      (0.0, 1.0, True, tf.float32),
      (0.0, 1.0, False, tf.float32),
      (0.0, 0.0, True, tf.float32),
      (0.0, 0.0, False, tf.float32),
      (1.0, 1.0, True, tf.float64),
      (1.0, 1.0, False, tf.float64),
      (1.0, 0.0, True, tf.float64),
      (1.0, 0.0, False, tf.float64),
      (0.0, 1.0, True, tf.float64),
      (0.0, 1.0, False, tf.float64),
      (0.0, 0.0, True, tf.float64),
      (0.0, 0.0, False, tf.float64)))
  def testModes(self,
                just_half_wave_rectify: float,
                one_capacitor: float,
                graph_mode: bool,
                dtype: tf.DType):
    ihc_params = carfac.IHCParams()
    ihc_params.one_capacitor = tf.constant(one_capacitor)
    ihc_params.just_half_wave_rectify = tf.constant(just_half_wave_rectify)
    car_params = carfac.CARParams()
    car_params.erb_per_step = tf.constant(3.0)
    carfac_cell = carfac.CARFACCell(ihc_params=ihc_params,
                                    car_params=car_params,
                                    num_ears=3,
                                    dtype=dtype)
    carfac_layer = tf.keras.layers.RNN(carfac_cell)
    model = tf.keras.Sequential([carfac_layer])
    impulse: np.ndarray = np.zeros([1, 128, 3, 1], dtype=np.float32)
    impulse[:, 0, 0, :] = 1
    def compute(data):
      return model(data)
    o = tf.function(compute)(impulse) if graph_mode else compute(impulse)
    self.assertEqual(o.shape, tf.TensorShape((1, 3, 12, 1)))


if __name__ == '__main__':
  tf.test.main()
